vault-core/src/main/java/org/apache/jackrabbit/vault/packaging/registry/impl/FSPackageRegistry.java

```diff
@@ -0,0 +1,570 @@
+            try {
+
+                FileUtils.copyInputStreamToFile(pump, tempFile);
+            } catch (Exception e) {
+                String msg = "Stream could be read successfully.";
+                log.error(msg);
+                throw new IOException(msg, e);
+            }
+    
+            if (archive.getJcrRoot() == null) {
+                String msg = "Stream is not a content package. Missing 'jcr_root'.";
+                log.error(msg);
+                throw new IOException(msg);
+            }
+    
+            final MetaInf inf = archive.getMetaInf();
+            PackagePropertiesImpl props = new PackagePropertiesImpl() {
+                @Override
+                protected Properties getPropertiesMap() {
+                    return inf.getProperties();
+                }
+            };
+            PackageId pid = props.getId();
+    
+            // invalidate pid if path is unknown
+            if (pid == null) {
+                pid = createRandomPid();
+            }
+            if (!pid.isValid()) {
+                throw new IllegalArgumentException("Unable to create package. Illegal package name.");
+            }
+    
+            File oldPkgFile = getPackageFile(pid);
+            String path = oldPkgFile.exists() ? oldPkgFile.getAbsolutePath() : getInstallationPath(pid);
+            String name = Text.getName(path);
+    
+            // TODO: @suess : capture installation state
+    
+            if (oldPkgFile.exists()) {
+                if (replace) {
+                    oldPkgFile.delete();
+                } else {
+                    throw new PackageExistsException("Package already exists: " + pid).setId(pid);
+                }
+            }
+            
+            ZipVaultPackage pkg = new ZipVaultPackage(archive, true);
+            File pkgFile = getPackageFile(pid);
+            FileUtils.moveFile(tempFile, pkgFile);
+            dispatch(PackageEvent.Type.UPLOAD, pid, null);
+            return pkg;
+        } finally {
+            pump.close();
+        }
+
+    }
+
+    @Nonnull
+    @Override
+    public PackageId register(@Nonnull File file, boolean replace) throws IOException, PackageExistsException {
+        ZipVaultPackage pack = new ZipVaultPackage(file, false, true);
+        File pkgFile = getPackageFile(pack.getId());
+        if (pkgFile.exists()) {
+            if (replace) {
+                pkgFile.delete();
+            } else {
+                throw new PackageExistsException("Package already exists: " + pack.getId()).setId(pack.getId());
+            }
+        }
+        FileUtils.copyFile(file, pkgFile);
+        return pack.getId();
+    }
+
+    @Nonnull
+    @Override
+    public PackageId registerExternal(@Nonnull File file, boolean replace) throws IOException, PackageExistsException {
+        ZipVaultPackage pack = new ZipVaultPackage(file, false, true);
+        // TODO: @suess - capture reference in central location
```

Comment:
or maybe add a `<package-path>.ref` to where the .zip file would be and just record the reference to the external file.

or just have a `<package-path>.properties`  as mandatory file that contains installation state etc. it could contain a `fileReference` property, in case the `<package-path>.zip` does not exist.